doc/proxy-metrics.md

```diff
@@ -0,0 +1,69 @@
++++
+title = "Proxy Metrics"
+docpage = true
+[menu.docs]
+  parent = "proxy-metrics"
++++
+
+The Conduit proxy exposes metrics that describe the traffic flowing through the
+proxy.  These metrics are exposed in Prometheus format on the proxy control port
+(4190 by default) at the `/metrics` path.  The following metrics will be
+exposed:
+
+### `request_total`
+
+A counter of the number of requests the proxy has received.
+
+### `request_duration_ms`
```

Comment:
i believe if we already have a `request_duration_ms`, then we'll get `request_duration_ms_count` for free, making `request_total` redundant? are there any labels that would apply to `request_total` that would not apply to `request_duration_ms`?

Reply:
I think this is true.  But maybe keeping the `request_total` and `response_total` metrics is still a good idea even if they are technically redundant because they help with discoverability and the queries will be a bit more readable.  e.g. queries about success rate won't reference duration

Reply:
Furthermore, they're subtly different.

request_total should count the number of requests _initiated_. you won't have a duration until the request _completes_. you wouldn't be able to know if requests are in-flight without the total count.

Also, the count is absolute counter, whereas histograms are time-bounded, and so effectively expose a rated count.  Rates should be avoided for counts we care about, as they can be lossy.

Reply:
I'm fine keeping them separate for naming purposes. This is an optimization we can keep in our back pocket if we need to reduce storage pressure on Prometheus.

To clarify a couple points:

> you wouldn't be able to know if requests are in-flight without the total count.

I think to measure in-flight correctly, we'd want a gauge. With our proposed implementation, I don't think doing `request_total - request_duration_ms_count` would be provide an accurate in-flight count, as collection time and counter increment ordering can occur in a few different ways.

> Also, the count is absolute counter

Histogram counters are absolute. It is only at query time where we can optionally provide a time-bound:
https://prometheus.io/docs/practices/histograms/#count-and-sum-of-observations

Reply:
I think `responses_total` actually has to be incremented when the response is complete, not when it is initiated, because it needs access to the grpc_status label value from the trailers.  For consistency, I think we should also increment `requests_total` when the request is complete.

Perhaps it also makes sense to have `requests_initiated_total` and `responses_initiated_total` counters but I'd prefer we get the basics in place first before deciding if we need this differentiation.